Applications/Spire/Include/Spire/Ui/AssociativeValueModel.hpp

```diff
@@ -0,0 +1,232 @@
+       */
+      std::shared_ptr<BooleanModel> find(const Type& value) const;
+
+      /**
+       * Returns Acceptable if the model's current value is valid, Invalid
+       * otherwise.
+       * @return The state of the AssociativeValueModel.
+       */
+      QValidator::State get_state() const override;
+
+      /**
+       * Returns the AssociativeValueModel's current value.
+       * @return The current value.
+       */
+      const Type& get_current() const override;
+
+      /**
+       * Sets the current value, iff a model is associated with the value.
+       * @param value The current value.
+       * @return Acceptable if the value was set successfully, Invalid
+       *         otherwise.
+       */
+      QValidator::State set_current(const Type& value) override;
+
+      /** Connects a slot to the current signal. */
+      boost::signals2::connection connect_current_signal(
+        const typename CurrentSignal::slot_type& slot) const override;
+
+    private:
+      mutable CurrentSignal m_current_signal;
+      Type m_current;
+      boost::optional<Type> m_default_value;
+      std::unordered_map<Type, std::shared_ptr<BooleanModel>> m_models;
+      std::vector<boost::signals2::scoped_connection> m_connections;
+      bool m_is_blocked;
+
+      void set_associated_model_value(const Type& value, bool model_value);
+      void on_current(const Type& value, bool is_selected);
+  };
+
+  template<typename T>
+  AssociativeValueModel<T>::AssociativeValueModel()
+    : m_is_blocked(false) {}
+
+  template<typename T>
+  AssociativeValueModel<T>::AssociativeValueModel(Type default_value)
+    : m_is_blocked(false),
+      m_default_value(std::move(default_value)) {}
+
+  template<typename T>
+  void AssociativeValueModel<T>::associate(
+      const std::shared_ptr<BooleanModel>& model, const Type& value) {
+    auto inserted_model = m_models.insert({value, model});
+    if(!inserted_model.second) {
+      return;
+    }
+    if(model->get_current()) {
+      model->set_current(false);
+    }
+    m_connections.push_back(model->connect_current_signal(
+      [=] (auto is_selected) { on_current(value, is_selected); }));
+    if(get_state() == QValidator::Invalid) {
+      set_current(value);
+    }
+  }
+
+  template<typename T>
+  std::shared_ptr<BooleanModel> AssociativeValueModel<T>::make_association(
+      const Type& value) {
+    auto model = std::make_shared<LocalBooleanModel>(false);
+    associate(model, value);
+    return model;
+  }
+
+  template<typename T>
+  std::shared_ptr<BooleanModel> AssociativeValueModel<T>::find(
+      const Type& value) const {
+    auto iterator = m_models.find(value);
+    if(iterator == m_models.end()) {
+      return nullptr;
+    }
+    return iterator->second;
+  }
+
+  template<typename T>
+  QValidator::State AssociativeValueModel<T>::get_state() const {
+    if(m_models.find(m_current) != m_models.end()) {
+      return QValidator::Acceptable;
+    }
+    return QValidator::Invalid;
```

Comment:
In fact, take it one step further, the model used in the constructor is the "default" model so that if every other model is false, then it is set to true. With this property, the ```optional<T>``` behavior comes for free.

Reply:
One caveat to this is that every AssociativeValueModel will have a default value. This might be fine for our purposes, but it's my understanding in the context of a radio button group that if the current selection is un-selected then it's effectively a no-op (i.e., the un-selected button is immediately reselected). If every model has a default we won't be able to model that behavior.

Reply:
There's no way to unselect a radio button group.

Reply:
My interpretation of the default value is that it becomes current if the current value is unselected, which creates problems with the typical button group use case. With the updated constructors the flow of the model would look like:

1. An AssociativeValueModel<string> is default constructed, resulting in an empty string becoming the default value.
2. BooleanModels are associated with "B" and "C".
3. The model for "C" is set to true.
4. The model associated with the default value becomes false in response.
5. The model for "C" is set to false (either unselected or something set the model to false).
6. The model associated with the default value is set to true in response because there are no other models that are true.

With this default value being constructed for every model in some way (either the default or explicitly set), it's not possible to model a radio button group because there's some other value that becomes current when the current value is unselected. The associated BooleanModels are purely reactive, so there's no guarantee that the current value's model can't be set to false by some external component. If step 5 is reverted so that the current value can't become false then we can't use the default value to model a nullable button group.

Reply:
Your point about the default value/model being used when the current model is set to false is correct, it's just that radio buttons don't set things to false, they only set things to true, and it so happens that when one model is set to true, the previous model is set to false.

Reply:
Good thoughts, I like that line of thinking. I don't think there is an issue in this case though, a model must always have a current value, and someone can use an optional to represent a value of none. No other model is ever in an ```Invalid``` state, only ```Intermediate``` or ```Accepted```. I'd prefer the property of always having a current value and being in a valid state, and users of this model ensure that by either being explicit about which value is engaged, or accepting the default value.